### PR TITLE
Apply search filters on database search results

### PR DESCRIPTION
**Pull request by tiksn**
Summary of the changes (in less than 80 chars)

 * separated query composition into local function
 * calling that function for initial query creation and after query changes

Addresses https://github.com/loic-sharma/BaGet/issues/417

### CODE REVIEW

File: src/BaGet.Core/Search/DatabaseSearchService.cs [hunk 1]
```diff
@@ -178,32 +178,39 @@ public Task IndexAsync(Package package, CancellationToken cancellationToken)
             var frameworks = GetCompatibleFrameworksOrNull(framework);
             var search = (IQueryable<Package>)_context.Packages.Where(p => p.Listed);
 
-            if (!string.IsNullOrEmpty(query))
+            IQueryable<Package> ComposePackageQuery(IQueryable<Package> packageQuery)
```
Comment (loic-sharma) [inline src/BaGet.Core/Search/DatabaseSearchService.cs]: What do you think of naming this `AddSearchFilters`?

```suggestion
            IQueryable<Package> AddSearchFilters(IQueryable<Package> packageQuery)
```

Comment (loic-sharma) [reply src/BaGet.Core/Search/DatabaseSearchService.cs]: Nice idea using a local function, it's a very clean solution 👍

File: src/BaGet.Core/Search/DatabaseSearchService.cs [hunk 2]
```diff
@@ -178,32 +178,39 @@ public Task IndexAsync(Package package, CancellationToken cancellationToken)
...
             {
```
Comment (loic-sharma) [inline src/BaGet.Core/Search/DatabaseSearchService.cs]: We could add `packageQuery = packageQuery.Where(p => p.Listed);` here.

This lets you simplify line 179 to:

```csharp
var search = (IQueryable<Package>)_context.Packages;
```

And lines 237-239 to:

```csharp
var results = await AddSearchFilters(search).ToListAsync(cancellationToken);
```

File: src/BaGet.Core/Search/DatabaseSearchService.cs [hunk 3]
```diff
@@ -178,32 +178,39 @@ public Task IndexAsync(Package package, CancellationToken cancellationToken)
...
-                query = query.ToLower();
-                search = search.Where(p => p.Id.ToLower().Contains(query));
-            }
+                if (!string.IsNullOrEmpty(query))
+                {
+                    query = query.ToLower();
+                    packageQuery = packageQuery.Where(p => p.Id.ToLower().Contains(query));
```
Comment (loic-sharma) [inline src/BaGet.Core/Search/DatabaseSearchService.cs]: We only need to filter on `query` once, could you move lines 183-187 out of `ComposePackageQuery` (or `AddSearchFilters`)? I would make line 179 something like:

```csharp
var search = AddSearchFilters(_context.Packages);
if (!string.IsNullOrEmpty(query))
{
  query = query.ToLower();
  search = search.Where(p => p.Id.ToLower().Contains(query));
}

var packageIds = ...
```

**Comment by loic-sharma** [comment]
Thank you for opening this! I left some minor suggestions, I'll merge this in once you've replied 😊

File: src/BaGet.Core/Search/DatabaseSearchService.cs (hunk 2 above)
Comment (tiksn) [reply src/BaGet.Core/Search/DatabaseSearchService.cs]: I would like to change/update `search` before using it. Maybe someone later will add some more lines of code, and I would like to have this filtering in that search already. Please let me know if you agree with it.

**Comment by loic-sharma** [comment]
Your changes look good. Before I merge this in, I will run the following test plan:

1. Create and upload packages:
    1. Create a package with versions:
        * 4.0.0-prerelease+semver2
        * 3.0.0+semver2
        * 2.0.0-prerelease
        * 1.0.0
    1. Create a package with a single prerelease and semver2 version
    1. Create a package with a single prerelease version
    1. Create a package with a single semver2 version
    1. Create a package with a single "normal" version
1. For each database, check that expected package shows up for combinations of prerelease and semver2 filter:
    * [ ] SQLite
    * [ ] PostgreSQL
    * [ ] MySQL
    * [ ] SQL Server

**Comment by loic-sharma** [comment]
Thank you for fixing this!

**Pull request merged**